File: frontend/src/services/dashboardService.ts
import apiClient from '@/lib/apiClient';

export interface Stats {
  todayVisitors: number;
  weekVisitors: number;
  weekNewComments: number;
  totalSubscribers: number;
}

export interface PopularPost {
  id: string;
  title: string;
  views: number;
}

export type TimeRange = '2weeks' | '30days' | '90days';

export interface ChartData {
  date: string;
  visitors: number;
}

export const dashboardService = {
  async getStats(slug: string): Promise<Stats> {
    const { data } = await apiClient.get<Stats>(`/api/blogs/${slug}/dashboard/stats`);
    return data;
  },

  async getPopularPosts(slug: string): Promise<PopularPost[]> {
    const { data } = await apiClient.get<PopularPost[]>(
      `/api/blogs/${slug}/dashboard/popular-posts`
    );
    return data;
  },

  async getTrafficData(slug: string, timeRange: TimeRange): Promise<ChartData[]> {
    const { data } = await apiClient.get<ChartData[]>(`/api/blogs/${slug}/dashboard/traffic`, {
      params: { timeRange },
    });
    return data;
  },
};

File: backend/blog-api/src/main/kotlin/com/contentria/api/dashboard/service/DashboardService.kt
package com.contentria.api.dashboard.service

import com.contentria.api.dashboard.dto.DashboardStatsResponse
import com.contentria.api.dashboard.dto.PopularPostResponse
import com.contentria.api.dashboard.dto.TimeRange
import com.contentria.api.dashboard.dto.TrafficChartDataResponse
import org.springframework.stereotype.Service
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import kotlin.random.Random

@Service
class DashboardService {

    fun getStats(slug: String): DashboardStatsResponse {
        // Dummy data for demonstration purposes
        return DashboardStatsResponse(
            todayVisitors = 0,
            weekVisitors = 0,
            weekNewComments = 0,
            totalSubscribers = 0
        )
    }

    fun getPopularPosts(slug: String): List<PopularPostResponse> {
        return listOf(
            PopularPostResponse(id = "1", title = "Understanding Kotlin Coroutines", views = 0),
            PopularPostResponse(id = "2", title = "Spring Boot with Kotlin: A Comprehensive Guide", views = 0),
            PopularPostResponse(id = "3", title = "Building RESTful APIs with Ktor", views = 0)
        )
    }

    fun getTrafficData(slug: String, timeRange: TimeRange): List<TrafficChartDataResponse> {
        val today = LocalDate.now()

        return when (timeRange) {
            TimeRange.TWO_WEEKS -> (0..13).map { i ->
                val date = today.minusDays(i.toLong())
                TrafficChartDataResponse(
                    date = date.format(DateTimeFormatter.ofPattern("MM/dd")),
                    visitors = 0
                )
            }.reversed()

            TimeRange.THIRTY_DAYS -> (0..29).map { i ->
                val date = today.minusDays(i.toLong())
                TrafficChartDataResponse(
                    date = date.format(DateTimeFormatter.ofPattern("MM/dd")),
                    visitors = 0 // 20 ~ 120
                )
            }.reversed()

            TimeRange.NINETY_DAYS -> (0..11).map { i ->
                val month = today.minusMonths(i.toLong())
                TrafficChartDataResponse(
                    date = month.format(DateTimeFormatter.ofPattern("yyyy/MM")),
                    visitors = 0 // 50 ~ 250
                )
            }.reversed()
        }
    }
}